File: app-laravel/src/Order/Application/Commands/CancelOrder/CancelOrderHandler.php
<?php

declare(strict_types=1);

namespace Src\Order\Application\Commands\CancelOrder;

use Src\Order\Domain\Repositories\OrderRepositoryInterface;
use Src\Order\Domain\Specifications\OrderCanBeCancelledSpec;
use Src\Order\Domain\ValueObjects\OrderId;
use Src\Shared\Application\Bus\Command;
use Src\Shared\Application\Bus\CommandHandler;
use Src\Shared\Domain\Exceptions\DomainException;

/**
 * CANCEL ORDER HANDLER (CQRS + Specification Pattern)
 * =====================================================
 * CancelOrderCommand-ı emal edən handler.
 *
 * BU HANDLER-DA SPECIFICATION PATTERN İSTİFADƏ OLUNUR:
 * - OrderCanBeCancelledSpec sifarişin ləğv edilə bilib-bilməyəcəyini yoxlayır.
 * - Handler birbaşa if/else yazmır — Specification-ı çağırır.
 * - Bu SRP (Single Responsibility Principle) prinsipinə uyğundur:
 *   Handler koordinasiya edir, Specification qaydanı yoxlayır.
 *
 * AXIN:
 * CancelOrderCommand
 *   → CancelOrderHandler
 *     → Repository-dən Order-i tap
 *     → OrderCanBeCancelledSpec ilə yoxla
 *     → Order.cancel() çağır
 *     → Repository-yə saxla
 */
class CancelOrderHandler implements CommandHandler
{
    public function __construct(
        private readonly OrderRepositoryInterface $orderRepository,
        private readonly OrderCanBeCancelledSpec $cancelSpec,
    ) {}

    /**
     * Sifariş ləğv etmə əmrini emal et.
     *
     * @param Command $command CancelOrderCommand olmalıdır
     * @throws DomainException Sifariş tapılmadıqda və ya ləğv edilə bilmədikdə
     */
    public function handle(Command $command): void
    {
        /** @var CancelOrderCommand $command */

        // 1. Sifarişi DB-dən tap
        $order = $this->orderRepository->findById(
            OrderId::fromString($command->orderId()),
        );

        if ($order === null) {
            throw new DomainException("Sifariş tapılmadı: {$command->orderId()}");
        }

        // 2. Specification ilə yoxla — ləğv etmək olarmı?
        // Bu if/else əvəzinə Specification istifadə etməyə nümunədir:
        // ✗ if ($order->status()->isPending() || $order->status()->isConfirmed())
        // ✓ if ($this->cancelSpec->isSatisfiedBy($order))
        if (!$this->cancelSpec->isSatisfiedBy($order)) {
            throw new DomainException(
                "Sifariş ləğv edilə bilməz. Cari status: {$order->status()->value()}. " .
                "Yalnız 'pending' və ya 'confirmed' statusunda ləğv etmək olar."
            );
        }

        // 3. Sifarişi ləğv et — domain metodu çağırılır
        // Order::cancel() daxilində OrderCancelledEvent qeydə alınır
        $order->cancel($command->reason());

        // 4. DB-yə saxla — domain event-lər dispatch olunacaq
        $this->orderRepository->save($order);
    }
}

File: app-spring/src/main/java/az/ecommerce/order/application/command/cancelorder/CancelOrderHandler.java
package az.ecommerce.order.application.command.cancelorder;

import az.ecommerce.order.domain.Order;
import az.ecommerce.order.domain.repository.OrderRepository;
import az.ecommerce.order.domain.specification.OrderCanBeCancelledSpec;
import az.ecommerce.order.domain.valueobject.OrderId;
import az.ecommerce.shared.application.bus.CommandHandler;
import az.ecommerce.shared.domain.exception.DomainException;
import az.ecommerce.shared.domain.exception.EntityNotFoundException;
import az.ecommerce.shared.infrastructure.bus.EventDispatcher;
import org.springframework.stereotype.Service;

@Service
public class CancelOrderHandler implements CommandHandler<CancelOrderCommand, Void> {

    private final OrderRepository repository;
    private final EventDispatcher eventDispatcher;

    public CancelOrderHandler(OrderRepository repository, EventDispatcher eventDispatcher) {
        this.repository = repository;
        this.eventDispatcher = eventDispatcher;
    }

    @Override
    public Void handle(CancelOrderCommand cmd) {
        Order order = repository.findById(new OrderId(cmd.orderId()))
                .orElseThrow(() -> new EntityNotFoundException("Order", cmd.orderId().toString()));

        if (!new OrderCanBeCancelledSpec().isSatisfiedBy(order)) {
            throw new DomainException("Bu sifariş hazırkı statusda ləğv edilə bilməz: " + order.status());
        }

        order.cancel(cmd.reason() != null ? cmd.reason() : "İstifadəçi tərəfindən");
        repository.save(order);
        eventDispatcher.dispatchAll(order);
        return null;
    }
}
